Add VariableTensorId, store it in TensorTypeSet

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25597 Add VariableTensorId, store it in TensorTypeSet**
* #25308 Tensor type set

We now take advantage of the new bitset representation TensorTypeSet to store "Variable-ness" of a tensor directly in the dispatch key. We introduce a new thread local TensorTypeSet "excluded" and replace the previous thread local boolean with it; we no longer have to query `is_variable()` to do dispatch (I didn't delete `is_variable`, because there are still a lot of uses of it). The key change is in `dispatchTypeId`.

Knock-on effects:
* Because Variable is now a TensorTypeId, I can eliminate the out-of-line registration `registerVariableOp` for variables; instead, make the registrar take a TensorTypeId (instead of a Backend) and you just register under the Variable key.
* Tensors aren't really ever created with Variable information initialized correctly at the start; instead, a tensor "becomes" a Variable because we set its `autograd_meta_`. These setters now correctly setup invariants on the dispatch type set. The new invariant is that if `autograd_meta_ != nullptr`, then `type_set().has(TensorTypeId::VariableTensorId)`.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17265919](https://our.internmc.facebook.com/intern/diff/D17265919)

ezyang committed:
Add VariableTensorId, store it in TensorTypeSet

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Update on "Add VariableTensorId, store it in TensorTypeSet"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Update on "Add VariableTensorId, store it in TensorTypeSet"

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25597 Add VariableTensorId, store it in TensorTypeSet**
* #25308 Tensor type set

We now take advantage of the new bitset representation TensorTypeSet to store "Variable-ness" of a tensor directly in the dispatch key. We introduce a new thread local TensorTypeSet "excluded" and replace the previous thread local boolean with it. This key change is in `dispatchTypeId`.

Knock-on effects:
* Because Variable is now a TensorTypeId, I can eliminate the out-of-line registration `registerVariableOp` for variables; registrar takes a TensorTypeId (instead of a Backend) and you just register under the Variable key.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Negation braino on "Add VariableTensorId, store it in TensorTypeSet"

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25597 Add VariableTensorId, store it in TensorTypeSet**
* #25308 Tensor type set

We now take advantage of the new bitset representation TensorTypeSet to store "Variable-ness" of a tensor directly in the dispatch key. We introduce a new thread local TensorTypeSet "excluded" and replace the previous thread local boolean with it; we no longer have to query `is_variable()` to do dispatch (I didn't delete `is_variable`, because there are still a lot of uses of it). The key change is in `dispatchTypeId`.

Knock-on effects:
* Because Variable is now a TensorTypeId, I can eliminate the out-of-line registration `registerVariableOp` for variables; instead, make the registrar take a TensorTypeId (instead of a Backend) and you just register under the Variable key.
* Tensors aren't really ever created with Variable information initialized correctly at the start; instead, a tensor "becomes" a Variable because we set its `autograd_meta_`. These setters now correctly setup invariants on the dispatch type set. The new invariant is that if `autograd_meta_ != nullptr`, then `type_set().has(TensorTypeId::VariableTensorId)`.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Update on "Add VariableTensorId, store it in TensorTypeSet"

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25597 Add VariableTensorId, store it in TensorTypeSet**
* #25308 Tensor type set

We now take advantage of the new bitset representation TensorTypeSet to store "Variable-ness" of a tensor directly in the dispatch key. We introduce a new thread local TensorTypeSet "excluded" and replace the previous thread local boolean with it; we no longer have to query `is_variable()` to do dispatch (I didn't delete `is_variable`, because there are still a lot of uses of it). The key change is in `dispatchTypeId`.

Knock-on effects:
* Because Variable is now a TensorTypeId, I can eliminate the out-of-line registration `registerVariableOp` for variables; instead, make the registrar take a TensorTypeId (instead of a Backend) and you just register under the Variable key.
* Tensors aren't really ever created with Variable information initialized correctly at the start; instead, a tensor "becomes" a Variable because we set its `autograd_meta_`. These setters now correctly setup invariants on the dispatch type set. The new invariant is that if `autograd_meta_ != nullptr`, then `type_set().has(TensorTypeId::VariableTensorId)`.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Update on "Add VariableTensorId, store it in TensorTypeSet"

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25597 Add VariableTensorId, store it in TensorTypeSet**
* #25308 Tensor type set

We now take advantage of the new bitset representation TensorTypeSet to store "Variable-ness" of a tensor directly in the dispatch key. We introduce a new thread local TensorTypeSet "excluded" and replace the previous thread local boolean with it; we no longer have to query `is_variable()` to do dispatch (I didn't delete `is_variable`, because there are still a lot of uses of it). The key change is in `dispatchTypeId`.

Knock-on effects:
* Because Variable is now a TensorTypeId, I can eliminate the out-of-line registration `registerVariableOp` for variables; instead, make the registrar take a TensorTypeId (instead of a Backend) and you just register under the Variable key.
* Tensors aren't really ever created with Variable information initialized correctly at the start; instead, a tensor "becomes" a Variable because we set its `autograd_meta_`. These setters now correctly setup invariants on the dispatch type set. The new invariant is that if `autograd_meta_ != nullptr`, then `type_set().has(TensorTypeId::VariableTensorId)`.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented:
cc'ing @ailzhang as this breaks XLA

ezyang committed:
rebase on "Add VariableTensorId, store it in TensorTypeSet"

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25597 Add VariableTensorId, store it in TensorTypeSet**
* #25308 Tensor type set

We now take advantage of the new bitset representation TensorTypeSet to store "Variable-ness" of a tensor directly in the dispatch key. We introduce a new thread local TensorTypeSet "excluded" and replace the previous thread local boolean with it; we no longer have to query `is_variable()` to do dispatch (I didn't delete `is_variable`, because there are still a lot of uses of it). The key change is in `dispatchTypeId`.

Knock-on effects:
* Because Variable is now a TensorTypeId, I can eliminate the out-of-line registration `registerVariableOp` for variables; instead, make the registrar take a TensorTypeId (instead of a Backend) and you just register under the Variable key.
* Tensors aren't really ever created with Variable information initialized correctly at the start; instead, a tensor "becomes" a Variable because we set its `autograd_meta_`. These setters now correctly setup invariants on the dispatch type set. The new invariant is that if `autograd_meta_ != nullptr`, then `type_set().has(TensorTypeId::VariableTensorId)`.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ailzhang commented:
Thanks @ezyang ! I'll take a look and prepare a patch tmr. :D

ezyang committed:
Rebase on "Add VariableTensorId, store it in TensorTypeSet"

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25597 Add VariableTensorId, store it in TensorTypeSet**
* #25308 Tensor type set

We now take advantage of the new bitset representation TensorTypeSet to store "Variable-ness" of a tensor directly in the dispatch key. We introduce a new thread local TensorTypeSet "excluded" and replace the previous thread local boolean with it; we no longer have to query `is_variable()` to do dispatch (I didn't delete `is_variable`, because there are still a lot of uses of it). The key change is in `dispatchTypeId`.

Knock-on effects:
* Because Variable is now a TensorTypeId, I can eliminate the out-of-line registration `registerVariableOp` for variables; instead, make the registrar take a TensorTypeId (instead of a Backend) and you just register under the Variable key.
* Tensors aren't really ever created with Variable information initialized correctly at the start; instead, a tensor "becomes" a Variable because we set its `autograd_meta_`. These setters now correctly setup invariants on the dispatch type set. The new invariant is that if `autograd_meta_ != nullptr`, then `type_set().has(TensorTypeId::VariableTensorId)`.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Update on "Add VariableTensorId, store it in TensorTypeSet"

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25597 Add VariableTensorId, store it in TensorTypeSet**
* #25308 Tensor type set

We now take advantage of the new bitset representation TensorTypeSet to store "Variable-ness" of a tensor directly in the dispatch key. We introduce a new thread local TensorTypeSet "excluded" and replace the previous thread local boolean with it; we no longer have to query `is_variable()` to do dispatch (I didn't delete `is_variable`, because there are still a lot of uses of it). The key change is in `dispatchTypeId`.

Knock-on effects:
* Because Variable is now a TensorTypeId, I can eliminate the out-of-line registration `registerVariableOp` for variables; instead, make the registrar take a TensorTypeId (instead of a Backend) and you just register under the Variable key.
* Tensors aren't really ever created with Variable information initialized correctly at the start; instead, a tensor "becomes" a Variable because we set its `autograd_meta_`. These setters now correctly setup invariants on the dispatch type set. The new invariant is that if `autograd_meta_ != nullptr`, then `type_set().has(TensorTypeId::VariableTensorId)`.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Update on "Add VariableTensorId, store it in TensorTypeSet"

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25597 Add VariableTensorId, store it in TensorTypeSet**
* #25308 Tensor type set

We now take advantage of the new bitset representation TensorTypeSet to store "Variable-ness" of a tensor directly in the dispatch key. We introduce a new thread local TensorTypeSet "excluded" and replace the previous thread local boolean with it; we no longer have to query `is_variable()` to do dispatch (I didn't delete `is_variable`, because there are still a lot of uses of it). The key change is in `dispatchTypeId`.

Knock-on effects:
* Because Variable is now a TensorTypeId, I can eliminate the out-of-line registration `registerVariableOp` for variables; instead, make the registrar take a TensorTypeId (instead of a Backend) and you just register under the Variable key.
* Tensors aren't really ever created with Variable information initialized correctly at the start; instead, a tensor "becomes" a Variable because we set its `autograd_meta_`. These setters now correctly setup invariants on the dispatch type set. The new invariant is that if `autograd_meta_ != nullptr`, then `type_set().has(TensorTypeId::VariableTensorId)`.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

bwasti commented on aten/src/ATen/core/ATenDispatch.h:
```diff
@@ -44,70 +46,54 @@ class CAFFE2_API ATenOpTable {
-    TORCH_CHECK(variable_function_ == nullptr,
+  void registerOp(TensorTypeId tid, void* fn) {
+    TORCH_CHECK(function_table_[static_cast<int64_t>(tid)] == nullptr,
         "Attempting to register variable function for schema ", schema_,
```
it's no longer necessarily "variable function"

bwasti commented on c10/core/TensorTypeSet.h:
```diff
@@ -33,7 +33,19 @@ namespace c10 {
+    : repr_(-1) {}
+  // Public version of TensorTypeSet(uint64_t) API; external users
+  // must be explicit when they do this!
+  TensorTypeSet(Raw, uint64_t x)
```
interesting way to distinguish this, why not `static from_raw` or something?

bwasti replied:
actually, this seems more discoverable, I like it

ezyang replied:
Also, sometimes you can use a constructor where you cannot use a static method.

bwasti commented on c10/core/impl/LocalTensorTypeSet.h:
```diff
@@ -0,0 +1,21 @@
+
+// TLS management for TensorTypeSet
+//
+// This manages thread-local TensorTypeSet of excluded keys which.  Keys which are
```
"excluded keys which" ?

bwasti reviewed (approved):
nits -- but generally this seems like quite nice cleanup

ezyang committed:
Update on "Add VariableTensorId, store it in TensorTypeSet"

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25597 Add VariableTensorId, store it in TensorTypeSet**
* #25308 Tensor type set

We now take advantage of the new bitset representation TensorTypeSet to store "Variable-ness" of a tensor directly in the dispatch key. We introduce a new thread local TensorTypeSet "excluded" and replace the previous thread local boolean with it; we no longer have to query `is_variable()` to do dispatch (I didn't delete `is_variable`, because there are still a lot of uses of it). The key change is in `dispatchTypeId`.

Knock-on effects:
* Because Variable is now a TensorTypeId, I can eliminate the out-of-line registration `registerVariableOp` for variables; instead, make the registrar take a TensorTypeId (instead of a Backend) and you just register under the Variable key.
* Tensors aren't really ever created with Variable information initialized correctly at the start; instead, a tensor "becomes" a Variable because we set its `autograd_meta_`. These setters now correctly setup invariants on the dispatch type set. The new invariant is that if `autograd_meta_ != nullptr`, then `type_set().has(TensorTypeId::VariableTensorId)`.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

gchanan commented on c10/core/impl/LocalTensorTypeSet.h:
```diff
@@ -0,0 +1,22 @@
+#include <c10/core/TensorTypeSet.h>
+
+// TLS management for TensorTypeSet
```
any reason you didn't call this TLSTensorTypeSet?  I feel like if that's the first comment, it should probably be in the name.

ezyang replied:
Because it is not "Thread local state tensor type set", it is "thread local tensor type set". But ThreadLocalTensorTypeSet is long so I dropped the Thread. I can rename it if you want.

gchanan commented on c10/core/TensorTypeId.h:
```diff
@@ -40,7 +40,11 @@ enum class TensorTypeId : uint8_t {
   ComplexCUDATensorId, // PyTorch only
 
-  // VariableTensorId,  // upcoming!
+  // WARNING! If you add more non-backend-like tensor ids (like tracing or
```
can we not use the word backend? :).

I admittedly don't have a better suggestion, but it seems the distinguishing characteristic is if you can write is a dispatch key in native_functions.yaml.  Speaking of which -- what happens if you try to do that now with Variable?

ezyang replied:
I think it will work, actually. (Not very well; you'll end up with a double registration, probably, in that case.)

gchanan commented on aten/src/ATen/core/ATenDispatch.h:
```diff
@@ -44,70 +46,54 @@ class CAFFE2_API ATenOpTable {
-    op_tables_.at(schema).registerVariableOp(reinterpret_cast<void*>(fn));
-    return *this;
+  template<class FuncType>
+  ATenDispatch& registerOp(Backend b, const char* schema, FuncType* fn) {
```
how much is this still used? :).

ezyang replied:
I don't know. Will have to remove and see, I guess!

ezyang replied:
Answer: `aten/src/ATen/test/extension_backend_test.cpp`

gchanan commented on c10/core/TensorImpl.h:
```diff
@@ -826,6 +826,11 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
    */
   void set_autograd_meta(std::unique_ptr<c10::AutogradMetaInterface> autograd_meta) {
     autograd_meta_ = std::move(autograd_meta);
+    if (autograd_meta_) {
```
this seems like the wrong layer of abstraction to do this?  Meaning, you are keeping VariableTensorId in sync with autograd_meta (which is reasonable), but I think these are all called by more-semantically-meaningful functions like "make_variable" or similar (and it seems fine to make this an error to be called from any other place).  I could be wrong and this would make it overly complicated, it depends on the implementation complexity.

ezyang replied:
Sort of. I agree with your general sentiment here, but at the end of the day, `autograd_meta_` and `type_set_` are private members on TensorImpl, and so methods on TensorImpl are the best place to ensure that the invariant is upheld, since it's easy to audit all of the sites which access these fields. It's probably reasonable to try to enforce these invariants at a higher level, but my preference would be that we'd start migrating some variable functionality into TensorImpl itself.

ezyang committed:
Update on "Add VariableTensorId, store it in TensorTypeSet"

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25597 Add VariableTensorId, store it in TensorTypeSet**
* #25308 Tensor type set

We now take advantage of the new bitset representation TensorTypeSet to store "Variable-ness" of a tensor directly in the dispatch key. We introduce a new thread local TensorTypeSet "excluded" and replace the previous thread local boolean with it; we no longer have to query `is_variable()` to do dispatch (I didn't delete `is_variable`, because there are still a lot of uses of it). The key change is in `dispatchTypeId`.

Knock-on effects:
* Because Variable is now a TensorTypeId, I can eliminate the out-of-line registration `registerVariableOp` for variables; instead, make the registrar take a TensorTypeId (instead of a Backend) and you just register under the Variable key.
* Tensors aren't really ever created with Variable information initialized correctly at the start; instead, a tensor "becomes" a Variable because we set its `autograd_meta_`. These setters now correctly setup invariants on the dispatch type set. The new invariant is that if `autograd_meta_ != nullptr`, then `type_set().has(TensorTypeId::VariableTensorId)`.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17265919](https://our.internmc.facebook.com/intern/diff/D17265919)

ezyang committed:
Update on "Add VariableTensorId, store it in TensorTypeSet"

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25597 Add VariableTensorId, store it in TensorTypeSet**
* #25308 Tensor type set

We now take advantage of the new bitset representation TensorTypeSet to store "Variable-ness" of a tensor directly in the dispatch key. We introduce a new thread local TensorTypeSet "excluded" and replace the previous thread local boolean with it; we no longer have to query `is_variable()` to do dispatch (I didn't delete `is_variable`, because there are still a lot of uses of it). The key change is in `dispatchTypeId`.

Knock-on effects:
* Because Variable is now a TensorTypeId, I can eliminate the out-of-line registration `registerVariableOp` for variables; instead, make the registrar take a TensorTypeId (instead of a Backend) and you just register under the Variable key.
* Tensors aren't really ever created with Variable information initialized correctly at the start; instead, a tensor "becomes" a Variable because we set its `autograd_meta_`. These setters now correctly setup invariants on the dispatch type set. The new invariant is that if `autograd_meta_ != nullptr`, then `type_set().has(TensorTypeId::VariableTensorId)`.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17265919](https://our.internmc.facebook.com/intern/diff/D17265919)

ezyang committed:
Update on "Add VariableTensorId, store it in TensorTypeSet"

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25597 Add VariableTensorId, store it in TensorTypeSet**
* #25308 Tensor type set

We now take advantage of the new bitset representation TensorTypeSet to store "Variable-ness" of a tensor directly in the dispatch key. We introduce a new thread local TensorTypeSet "excluded" and replace the previous thread local boolean with it; we no longer have to query `is_variable()` to do dispatch (I didn't delete `is_variable`, because there are still a lot of uses of it). The key change is in `dispatchTypeId`.

Knock-on effects:
* Because Variable is now a TensorTypeId, I can eliminate the out-of-line registration `registerVariableOp` for variables; instead, make the registrar take a TensorTypeId (instead of a Backend) and you just register under the Variable key.
* Tensors aren't really ever created with Variable information initialized correctly at the start; instead, a tensor "becomes" a Variable because we set its `autograd_meta_`. These setters now correctly setup invariants on the dispatch type set. The new invariant is that if `autograd_meta_ != nullptr`, then `type_set().has(TensorTypeId::VariableTensorId)`.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17265919](https://our.internmc.facebook.com/intern/diff/D17265919)

zdevito reviewed (approved):
This looks good to me!

ezyang commented on c10/core/TensorImpl.cpp:
```diff
@@ -57,7 +58,7 @@ TensorImpl::TensorImpl(Storage&& storage, TensorTypeSet type_set, const caffe2::
       data_type_(data_type),
       device_opt_(device_opt),
-      type_set_(type_set) {
+      type_set_(type_set.remove(TensorTypeId::VariableTensorId)) {
```
Yeah, it's pretty goofy. Basically, the invariant is `type_set.has(TensorTypeId::VariableTensorId) iff autograd_meta_ != nullptr`, right? Well, on construction, `autograd_meta_` is always null. So to preserve the invariant we have to remove variable tensor id if it was passed in (which it might be, if you are doing a shallow copy from a variable thing).

@gchanan mentioned earlier that we could make things a bit better if we moved some of the type set logic into the variable constructors. If we can move `autograd_meta_` initialization into these constructors, we can make this a bit more transparent. But that involves a lot more moving code around. In any case, we probably need a Note for this, I'll go add one.

zdevito replied:
Yeah, I think it is fine, especially for now. Ideally this would assert if variable was set, and we'd take care not to shallow initialize something with variable without also setting autograd_meta_ at the same time. I think this is similar or equivalent to what you were saying in your comment.
